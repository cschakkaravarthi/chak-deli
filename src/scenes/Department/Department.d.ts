import React, { FC } from 'react';
import { Department as DepartmentModel } from '../../shared/models/Department.model';
import { match } from 'react-router';
declare type Props = {
    match?: match;
    departmentId?: number;
    department?: DepartmentModel;
    fetchDepartmentById?: (departmentId: number) => Promise<void>;
};
export declare const Department: FC<Props>;
declare const _default: import("react-redux").ConnectedComponent<React.FC<Props>, Pick<Props, "match" | "departmentId" | "department">>;
export default _default;
//# sourceMappingURL=Department.d.ts.map