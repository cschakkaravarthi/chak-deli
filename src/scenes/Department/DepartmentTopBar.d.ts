import React, { FC } from 'react';
import { Department as DepartmentModel } from '../../shared/models/Department.model';
import { ContentCategorizedStorageModel, VariousContentModel } from '../../shared/types/contentTypes';
import { Page } from '../../shared/models/Page.model';
declare type Props = {
    content?: ContentCategorizedStorageModel;
    department?: DepartmentModel;
    page?: Page;
    departmentId: number;
    active?: number;
    fetchContent?: (filterId: string) => void;
    fetchDepartmentById?: (departmentId: number) => void;
    navItems?: VariousContentModel[];
};
export declare const DepartmentTopBar: FC<Props>;
declare const _default: import("react-redux").ConnectedComponent<React.FC<Props>, Pick<Props, "departmentId" | "department" | "page" | "content" | "active" | "navItems">>;
export default _default;
//# sourceMappingURL=DepartmentTopBar.d.ts.map