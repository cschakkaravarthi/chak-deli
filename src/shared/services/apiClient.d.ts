declare const client: import("axios").AxiosInstance;
export default client;
//# sourceMappingURL=apiClient.d.ts.map