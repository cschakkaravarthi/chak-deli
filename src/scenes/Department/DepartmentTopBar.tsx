import React, { FC, useEffect } from 'react';
import { LogoBanner } from 'umgc_ui_library';
import images from '../../images/images';
import Container from 'react-bootstrap/Container';
import TopNavMinimal, {
  LinkInterface
} from '../../components/nav/TopNavMinimal';
import { DepartmentTemplate, Department as DepartmentModel } from '../../shared/models/Department.model';
import Breadcrumb from '../../components/nav/Breadcrumb';
import { AppState } from '../../shared/types/genericTypes';
import { connect } from 'react-redux';
import { fetchDepartmentById } from '../../shared/actions/departmentActions';
import { fetchContent } from '../../shared/actions/contentActions';
import get from 'lodash.get';
import {
  ContentQueryParamsType, FacetedContentModel,
  VariousContentModel
} from '../../shared/types/contentTypes';
import { getLinkToContent, toQueryString } from '../../utils/content';
import { Page } from '../../shared/models/Page.model';
import DrupalIds from '../../drupalIds';
import { Link } from 'react-router-dom';

type Props = {
  active?: number;
  content?: FacetedContentModel;
  department?: VariousContentModel;
  departmentId: number;
  fetchContent?: (params: ContentQueryParamsType) => void;
  fetchDepartmentById?: (departmentId: number) => void;
  navItems?: VariousContentModel[];
  page?: Page;
  paramSet?: ContentQueryParamsType;
};

const mapState: any = (state: AppState, props: Props): Props => {
  const departmentIds = get(DrupalIds.departments, props.departmentId);
  const paramSet: ContentQueryParamsType = {
    category: departmentIds && departmentIds.topNav
  };
  const topNavContent = get(state.contentReducers.content, `[${toQueryString(paramSet)}]`);
  return {
    ...props,
    paramSet,
    content: topNavContent,
    department: state.departmentReducers.departments.content.find(
      d => d.drupal_id === props.departmentId
    )
  };
};

const actionCreators = { fetchContent, fetchDepartmentById };

const sortAlphabetically = (links: LinkInterface[]): LinkInterface[] => {
  return links.sort((a: LinkInterface, b: LinkInterface) => {
    if (a.title < b.title) {
      return -1;
    }
    if (a.title > b.title) {
      return 1;
    }
    return 0;
  });
};

const getNavigationLinks = (
  department: VariousContentModel,
  topNavContent: FacetedContentModel,
  active?: number,
  containsPeople = false
): LinkInterface[] => {
  const navLinks = [];
  // First we add the department to the links (not active because we are in a page component)
  navLinks.push({
    title: 'Home',
    isActive: active === department.drupal_id,
    id: department.drupal_id,
    linkComponent: getLinkToContent(department, 'Home')
  });

  const pageNavLinks: LinkInterface[] = [];

  // Then, we add the pages as links (set active the one that matches the current page)
  topNavContent.content.map(c => {
    pageNavLinks.push({
      isActive: active === c.drupal_id,
      title: c.title,
      id: c.drupal_id,
      linkComponent: getLinkToContent(c, c.title)
    });
  });

  const sortedPageNavLinks: LinkInterface[] = sortAlphabetically(pageNavLinks);

  navLinks.push(...sortedPageNavLinks);

  if (containsPeople) {
    navLinks.push({
      id: 0,
      isActive: active === -1,
      title: 'People',
      linkComponent: (p: any) => (
        <Link to={'/departments/' + department.drupal_id + '/people'} {...p}>People</Link>
      )
    });
  }

  return navLinks;
};

export const DepartmentTopBar: FC<Props> = props => {
  const {
    active,
    content,
    department,
    departmentId,
    fetchContent,
    fetchDepartmentById,
    page,
    paramSet
  } = props;

  useEffect(() => {
    fetchDepartmentById && fetchDepartmentById(departmentId);
  }, [departmentId]);

  useEffect(() => {
    fetchContent && paramSet && fetchContent(paramSet);
  }, []);

  const getTemplateId = (): string => {
    const getTemplates = get(
      department,
      'template',
      [] as DepartmentTemplate[]
    );
    // Typecast to string.
    return getTemplates[0].drupal_id + '';
  };

  const templateId = department && getTemplateId();

  const wrappedDepartment = department as DepartmentModel;

  const containsPeople = wrappedDepartment && Array.isArray(wrappedDepartment.contacts) && wrappedDepartment.contacts.length > 0 && templateId !== DrupalIds.templates.templateE && templateId !== DrupalIds.templates.templateB;

  const navLinks =
    department && content
      ? getNavigationLinks(department, content, active, containsPeople)
      : [];

  if (!department) {
    return null;
  }

  return (
    <>
      <LogoBanner
        logoSrc={images.logoSrc}
        bannerSrc={get(department, 'banner_uri.umgc_banner', '')}
      />
      {content && !!content.content.length && (
        <TopNavMinimal links={navLinks} />
      )}
      <Container>
        {page ? (
          <Breadcrumb
            variant="secondary"
            subsection={page.title}
            title={department.title}
            sectionImageSrc={get(
              department,
              'logo_uri',
              require('../../assets/default_icon_circle.svg')
            )}
          />
        ) : (
          <Breadcrumb
            title={department.title}
            bodyText={department.summary}
            sectionImageSrc={get(
              department,
              'logo_uri',
              require('../../assets/default_icon_circle.svg')
            )}
          />
        )}
      </Container>
    </>
  );
};

export default connect(mapState, actionCreators)(DepartmentTopBar);
