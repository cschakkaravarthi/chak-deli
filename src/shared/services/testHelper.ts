import { ReactElement } from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';

import getStore from './mockGlobalStore';

export const shallowRender = (toRender: JSX.Element): ReactElement => {
  const renderer = ShallowRenderer.createRenderer();
  renderer.render(toRender);
  return renderer.getRenderOutput();
};

export const getMockStore = getStore;

export const drawerMockNavigation = {
  state: {
    isDrawerOpen: true,
    isTransitioning: false,
    index: 1,
    routes: [],
    key: 'abc',
    params: {}
  },
  dispatch: jest.fn(),
  goBack: jest.fn(),
  dismiss: jest.fn(),
  navigate: jest.fn(),
  openDrawer: jest.fn(),
  closeDrawer: jest.fn(),
  toggleDrawer: jest.fn(),
  getParam: jest.fn(),
  setParams: jest.fn(),
  addListener: jest.fn(),
  push: jest.fn(),
  replace: jest.fn(),
  reset: jest.fn(),
  pop: jest.fn(),
  popToTop: jest.fn(),
  isFocused: jest.fn(),
  isFirstRouteInParent: jest.fn(),
  dangerouslyGetParent: jest.fn()
};
