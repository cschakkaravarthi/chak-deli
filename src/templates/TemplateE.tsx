import React, { FC, useEffect } from 'react';
import Col from 'react-bootstrap/Col';
import Row from 'react-bootstrap/Row';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import { Department } from '../shared/models/Department.model';
import UserInfoModel from '../shared/models/UserInfo.model';
import { TemplateIdsModel } from '../shared/models/Template.model';
import { fetchUsersByEmail } from '../shared/actions/usersActions';
import { AppState } from '../shared/types/genericTypes';
import createDOMPurify from 'dompurify';
import get from 'lodash.get';
import find from 'lodash.find';
import { Card, ContentWrapper } from 'umgc_ui_library/lib';
import LinksCards from '../components/links/LinksCards';

type Props = {
  content: Department;
  contentIds: TemplateIdsModel;
  departmentContacts?: UserInfoModel[];
  fetchUsersByEmail?: (email: string) => void;
};

const mapState: any = (state: AppState, props: Props): Props => {
  return {
    ...props,
    departmentContacts: state.usersReducers.users
  };
};

const actionCreators = { fetchUsersByEmail };

export const Template: FC<Props> = props => {
  const { content, departmentContacts, fetchUsersByEmail } = props;
  const flatContacts = content.contacts
    ? content.contacts.reduce((acc: string[], val) => acc.concat(val.contacts), [])
    : [];

  useEffect(() => {
    fetchUsersByEmail && fetchUsersByEmail(flatContacts.join(','));
  }, []);

  return (
    <>
      {props.content.body && (
        <Row className="px-3 pt-3">
          <ContentWrapper title='' topBorder="darkTurquoise">
            <div
              dangerouslySetInnerHTML={{
                __html: createDOMPurify.sanitize(props.content.body)
              }}
            />
          </ContentWrapper>
        </Row>
      )}

      {!!flatContacts.length && props.contentIds.contentBuckets[0] && (
        <Row>
          <Col md={6} className="py-md-5 py-2">
            <LinksCards category={props.contentIds.contentBuckets[0]} />
          </Col>
          <Col md={6} className="py-md-5 py-2">
            <Card
              title="Contacts"
              topBorder="green"
              variant="standard"
              type="cardsInContainer"
            >
              <Row>
                {
                  departmentContacts && departmentContacts.map(c => {
                    const officeNumber = find(c.phone, { type: 'desk' });
                    const mobileNumber = find(c.phone, { type: 'mobile' });
                    return (
                      <Col key={c.email} md={12} className="p-1">
                        <Card {...{
                          type: 'contactMe',
                          variant: 'profile',
                          title: get(c, 'fullName', ''),
                          office: get(officeNumber, 'number', ''),
                          mobile: get(mobileNumber, 'number', ''),
                          email: c.email,
                          contactPic: get(c, 'profilePicture.medium'),
                          isCurrentUser: false,
                          LinkToProfile: (p: React.PropsWithChildren<{}>) => (
                            <Link to={`/profile?upn=${c.email}`}>{p.children}</Link>
                          )
                        }} />
                      </Col>
                    );
                  })
                }
              </Row>
            </Card>
          </Col>
        </Row>
      )}
    </>
  );
};

export default connect(mapState, actionCreators)(Template);
