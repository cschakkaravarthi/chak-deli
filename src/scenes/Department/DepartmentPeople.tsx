import React, { FC, useEffect } from 'react';
import Container from 'react-bootstrap/Container';
import Col from 'react-bootstrap/Col';
import Row from 'react-bootstrap/Row';
import {
  DepartmentTemplate,
  ContactGroup
} from '../../shared/models/Department.model';
import { match } from 'react-router';
import { AppState } from '../../shared/types/genericTypes';
import { connect } from 'react-redux';
import { fetchDepartmentById } from '../../shared/actions/departmentActions';
import get from 'lodash.get';
import find from 'lodash.find';
import { Page as PageModel } from '../../shared/models/Page.model';
import { Card } from 'umgc_ui_library/lib';
import { Link } from 'react-router-dom';
import { contentToDepartment, VariousContentModel } from '../../shared/types/contentTypes';
import UserInfoModel from '../../shared/models/UserInfo.model';
import DepartmentTopBar from './DepartmentTopBar';
import { fetchUsersByEmail } from '../../shared/actions/usersActions';

type Props = {
  match?: match;
  departmentId?: number;
  department?: VariousContentModel;
  departmentContacts?: UserInfoModel[];
  fetchDepartmentById?: (departmentId: number) => Promise<void>;
  fetchUsersByEmail?: (email: string) => void;
};

const mapState: any = (state: AppState, props: Props): Props => {
  const departmentId = parseInt(get(props, 'match.params.id'), 10);
  return {
    ...props,
    departmentId,
    departmentContacts: state.usersReducers.users,
    department: state.departmentReducers.departments.content.find(
      d => d.drupal_id === departmentId
    )
  };
};

const actionCreators = { fetchDepartmentById, fetchUsersByEmail };

const getPeoplePage = (): PageModel => {
  return {
    type: 'page',
    language: 'en',
    categories: [],
    category_ancestors: [],
    owner: [],
    drupal_id: 0,
    title: 'People'
  };
};

const getGroupContacts = (contactGroup: ContactGroup, allContacts: (UserInfoModel[])): UserInfoModel[] => {
  contactGroup.contacts = contactGroup.contacts.map(c => c.toLowerCase());
  return allContacts.filter(c => {
    if (c && c.email) {
      return contactGroup.contacts.includes(c.email.toLowerCase());
    }
    return false;
  });
};

export const DepartmentPeople: FC<Props> = props => {
  const {
    departmentId,
    department,
    fetchDepartmentById,
    fetchUsersByEmail,
    departmentContacts
  } = props;

  useEffect(() => {
    departmentId && fetchDepartmentById && fetchDepartmentById(departmentId);
  }, [departmentId, fetchDepartmentById]);

  let flatContacts: string[] = [];

  let typedDepartment = null;

  if (department) {
    typedDepartment = contentToDepartment(department);
    flatContacts = typedDepartment && typedDepartment.contacts
      ? typedDepartment.contacts.reduce((acc: string[], val) => acc.concat(val.contacts), [])
      : [];
  }

  useEffect(() => {
    if (flatContacts.length > 0) {
      fetchUsersByEmail && fetchUsersByEmail(flatContacts.join(','));
    }
  }, [department]);

  const getTemplateId = (): string => {
    const getTemplates = get(
      department,
      'template',
      [] as DepartmentTemplate[]
    );
    return getTemplates[0].drupal_id + '';
  };

  const templateId = department && getTemplateId();

  const page = getPeoplePage();

  return departmentId && department && typedDepartment && departmentContacts && templateId ? (
    <div className='departments pb-5'>
      <DepartmentTopBar
        departmentId={departmentId}
        active={-1}
        page={page}
      />
      <Container>
        {typedDepartment.contacts && typedDepartment.contacts.map(contactGroup => {
          const groupContacts = getGroupContacts(contactGroup, departmentContacts);
          return (
            <Row key={contactGroup.group_label}>
              <Col md={12} className='py-md-2 py-2'>
                <Card
                  title={contactGroup.group_label || 'Contacts'}
                  topBorder='green'
                  variant='standard'
                  type='cardsInContainer'>
                  <Row>
                    {groupContacts &&
                      groupContacts.map(c => {
                        const officeNumber = find(c.phone, { type: 'desk' });
                        const mobileNumber = find(c.phone, { type: 'mobile' });
                        return (
                          <Col key={c.email} md={6} className='py-1 px-3'>
                            <Card
                              {...{
                                type: 'contactMe',
                                variant: 'profile',
                                title: get(c, 'fullName', ''),
                                office: get(officeNumber, 'number', ''),
                                mobile: get(mobileNumber, 'number', ''),
                                email: c.email,
                                contactPic: get(c, 'profilePicture.medium'),
                                isCurrentUser: false,
                                LinkToProfile: (p: React.PropsWithChildren<{}>) => (
                                  <Link to={`/profile?upn=${c.email}`}>
                                    {p.children}
                                  </Link>
                                )
                              }}
                            />
                          </Col>
                        );
                      })}
                  </Row>
                </Card>
              </Col>
            </Row>
          );
        })}
      </Container>
    </div>
  ) : null;
};

export default connect(mapState, actionCreators)(DepartmentPeople);
