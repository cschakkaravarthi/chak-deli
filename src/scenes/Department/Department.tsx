import React, { FC, useEffect, useState } from 'react';
import Container from 'react-bootstrap/Container';
import TemplateA from '../../templates/TemplateA';
import TemplateB from '../../templates/TemplateB';
import TemplateC from '../../templates/TemplateC';
import TemplateD from '../../templates/TemplateD';
import TemplateE from '../../templates/TemplateE';
import {
  DepartmentTemplate
} from '../../shared/models/Department.model';
import { match } from 'react-router';
import { AppState } from '../../shared/types/genericTypes';
import { connect } from 'react-redux';
import { fetchDepartmentById } from '../../shared/actions/departmentActions';
import get from 'lodash.get';
import DrupalIds from '../../drupalIds';
import DepartmentTopBar from './DepartmentTopBar';
import Error404 from '../Errors/Error404';
import { contentToDepartment, VariousContentModel } from '../../shared/types/contentTypes';

type Props = {
  match?: match;
  departmentId?: number;
  department?: VariousContentModel;
  fetchDepartmentById?: (departmentId: number) => Promise<void>;
};

const mapState: any = (state: AppState, props: Props): Props => {
  const departmentId = parseInt(get(props, 'match.params.id'), 10);
  return {
    ...props,
    departmentId,
    department: state.departmentReducers.departments.content.find(
      d => d.drupal_id === departmentId
    )
  };
};

const actionCreators = { fetchDepartmentById };

export const Department: FC<Props> = props => {
  const { departmentId, department, fetchDepartmentById } = props;

  const [isFetching, setIsFetching] = useState<boolean>(true);

  if (!departmentId) {
    if (!isFetching) {
      return <Error404 />;
    } else {
      return null;
    }
  }

  const departmentIds = get(DrupalIds.departments, departmentId);

  useEffect(() => {
    if (fetchDepartmentById) {
      fetchDepartmentById(departmentId)
        .then(() => {
          setIsFetching(false);
        })
        .catch(() => {
          setIsFetching(false);
        });
    }
  }, []);

  if (!department) {
    if (!isFetching) {
      return <Error404 />;
    } else {
      return null;
    }
  }

  const getTemplateId = (): string => {
    const getTemplates = get(
      department,
      'template',
      [] as DepartmentTemplate[]
    );
    // Typecast to string.
    return getTemplates[0].drupal_id + '';
  };

  const templateId = department && getTemplateId();

  // Which template component should we use? To find out we check the template ID.
  // If we don't recognize it we default to TemplateA.
  const TemplateComponent =
    templateId === DrupalIds.templates.templateB
      ? TemplateB
      : templateId === DrupalIds.templates.templateC
        ? TemplateC
        : templateId === DrupalIds.templates.templateD
          ? TemplateD
          : templateId === DrupalIds.templates.templateE
            ? TemplateE
            : TemplateA;

  const typedDepartment = contentToDepartment(department);
  return typedDepartment && templateId ? (
    <div className="departments pb-5">
      <DepartmentTopBar departmentId={departmentId} active={departmentId} />
      <Container>
        <TemplateComponent content={typedDepartment} contentIds={departmentIds} />
      </Container>
    </div>
  ) : null;
};

export default connect(mapState, actionCreators)(Department);
