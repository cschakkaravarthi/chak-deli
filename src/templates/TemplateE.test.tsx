import React from 'react';
import { shallowRender } from '../shared/services/testHelper';
import { Template } from './TemplateE';
import { dummyDepartment, dummyDepartmentContacts } from '../shared/models/Department.model';
import { dummyTemplateIds } from '../shared/models/Template.model';

describe('TemplateE', () => {
  it('renders correctly', () => {
    const tree = shallowRender(
      <Template
        content={dummyDepartment}
        contentIds={dummyTemplateIds}
        departmentContacts={dummyDepartmentContacts}
      />
    );
    expect(tree).toMatchSnapshot();
  });
});
