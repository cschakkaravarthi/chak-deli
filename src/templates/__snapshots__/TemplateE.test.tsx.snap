// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TemplateE renders correctly 1`] = `
<React.Fragment>
  <ForwardRef
    className="px-3 pt-3"
    noGutters={false}
  >
    <Unknown
      title=""
      topBorder="darkTurquoise"
    >
      <div
        dangerouslySetInnerHTML={
          Object {
            "__html": "<p>Lorem ipsum dolor sit amet, consectetur adipiscing elit. Praesent vitae odio vehicula nisl lacinia luctus non et libero. Quisque leo tortor, lacinia et porttitor vel, dignissim at sapien.</p>",
          }
        }
      />
    </Unknown>
  </ForwardRef>
  <ForwardRef
    noGutters={false}
  >
    <ForwardRef
      className="py-md-5 py-2"
      md={6}
    >
      <UNDEFINED
        category="3"
      />
    </ForwardRef>
    <ForwardRef
      className="py-md-5 py-2"
      md={6}
    >
      <Unknown
        title="Contacts"
        topBorder="green"
        type="cardsInContainer"
        variant="standard"
      >
        <ForwardRef
          noGutters={false}
        >
          <ForwardRef
            className="p-1"
            md={12}
          >
            <Unknown
              LinkToProfile={[Function]}
              email="doris.aves@umusic.com"
              isCurrentUser={false}
              mobile=""
              office=""
              title="Luz Lacson"
              type="contactMe"
              variant="profile"
            />
          </ForwardRef>
        </ForwardRef>
      </Unknown>
    </ForwardRef>
  </ForwardRef>
</React.Fragment>
`;
